File: web/src/features/admin-categories/services/categories.ts
import { apiFetch } from '@/lib/api';
import type {
  AdminCategory,
  AdminCategoryCourseSummary,
  AdminCategoryCourseTranslation,
  AdminCategoryDetail,
  AdminCategoryDetailStats,
  AdminCategoryListResponse,
  AdminCategoryStats,
  AdminCategoryStatsResponse,
  CategoryTranslationFull,
  CreateCategoryPayload,
  ListCategoriesParams,
  UpdateCategoryPayload,
} from '../types';

export class CategoryApiError extends Error {
  code: string;
  status: number;

  constructor(code: string, status: number) {
    super(code);
    this.name = 'CategoryApiError';
    this.code = code;
    this.status = status;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function parseTranslation(value: unknown): { lang: string; name: string; description: string | null } {
  if (!isRecord(value)) {
    return { lang: '', name: '', description: null };
  }

  return {
    lang: typeof value.lang === 'string' ? value.lang : '',
    name: typeof value.name === 'string' ? value.name : '',
    description: typeof value.description === 'string' ? value.description : null,
  };
}

function parseTranslationFull(value: unknown): CategoryTranslationFull {
  const translation = parseTranslation(value);
  if (!isRecord(value)) {
    throw new Error('INVALID_CATEGORY_TRANSLATION_PAYLOAD');
  }

  return {
    ...translation,
    id: typeof value.id === 'string' ? value.id : '',
    category_id: typeof value.category_id === 'string' ? value.category_id : '',
    is_auto_translated: typeof value.is_auto_translated === 'boolean' ? value.is_auto_translated : false,
    created_at: typeof value.created_at === 'string' ? value.created_at : '',
    updated_at: typeof value.updated_at === 'string' ? value.updated_at : '',
  };
}

function parseCategory(value: unknown): AdminCategory {
  if (!isRecord(value)) {
    throw new Error('INVALID_CATEGORY_PAYLOAD');
  }

  const rawTranslations = isRecord(value.translations) ? value.translations : {};
  const translations = Object.entries(rawTranslations).reduce<Record<string, ReturnType<typeof parseTranslation>>>(
    (acc, [key, translation]) => {
      acc[key] = parseTranslation(translation);
      return acc;
    },
    {},
  );

  return {
    id: typeof value.id === 'string' ? value.id : '',
    icon: typeof value.icon === 'string' ? value.icon : null,
    created_at: typeof value.created_at === 'string' ? value.created_at : '',
    updated_at: typeof value.updated_at === 'string' ? value.updated_at : '',
    translations,
    course_count: typeof value.course_count === 'number' ? value.course_count : 0,
  };
}

function parseCourseTranslation(value: unknown): AdminCategoryCourseTranslation {
  if (!isRecord(value)) {
    return { lang: '', title: '', subtitle: null, description: null, slug: null };
  }

  return {
    lang: typeof value.lang === 'string' ? value.lang : '',
    title: typeof value.title === 'string' ? value.title : '',
    subtitle: typeof value.subtitle === 'string' ? value.subtitle : null,
    description: typeof value.description === 'string' ? value.description : null,
    slug: typeof value.slug === 'string' ? value.slug : null,
  };
}

function parseCourseSummary(value: unknown): AdminCategoryCourseSummary {
  if (!isRecord(value)) {
    throw new Error('INVALID_CATEGORY_COURSE_PAYLOAD');
  }

  const rawTranslations = isRecord(value.translations) ? value.translations : {};
  const translations = Object.entries(rawTranslations).reduce<Record<string, AdminCategoryCourseTranslation>>(
    (acc, [key, translation]) => {
      acc[key] = parseCourseTranslation(translation);
      return acc;
    },
    {},
  );

  return {
    id: typeof value.id === 'string' ? value.id : '',
    default_slug: typeof value.default_slug === 'string' ? value.default_slug : null,
    thumbnail_url: typeof value.thumbnail_url === 'string' ? value.thumbnail_url : null,
    level: typeof value.level === 'string' ? value.level : '',
    is_published: typeof value.is_published === 'boolean' ? value.is_published : false,
    approval_status: typeof value.approval_status === 'string' ? value.approval_status : null,
    base_price: typeof value.base_price === 'number' ? value.base_price : 0,
    currency: typeof value.currency === 'string' ? value.currency : 'VND',
    estimated_duration: typeof value.estimated_duration === 'number' ? value.estimated_duration : null,
    lessons_count: typeof value.lessons_count === 'number' ? value.lessons_count : 0,
    views: typeof value.views === 'number' ? value.views : 0,
    total_enrolls: typeof value.total_enrolls === 'number' ? value.total_enrolls : 0,
    revenue: typeof value.revenue === 'number' ? value.revenue : 0,
    rating_avg: typeof value.rating_avg === 'number' ? value.rating_avg : 0,
    created_at: typeof value.created_at === 'string' ? value.created_at : '',
    updated_at: typeof value.updated_at === 'string' ? value.updated_at : '',
    translations,
  };
}

function parseDetailStats(value: unknown): AdminCategoryDetailStats {
  if (!isRecord(value)) {
    return {
      total_courses: 0,
      published_courses: 0,
      draft_courses: 0,
      total_enrolls: 0,
      total_revenue: 0,
      total_views: 0,
      avg_rating: 0,
    };
  }

  return {
    total_courses: typeof value.total_courses === 'number' ? value.total_courses : 0,
    published_courses: typeof value.published_courses === 'number' ? value.published_courses : 0,
    draft_courses: typeof value.draft_courses === 'number' ? value.draft_courses : 0,
    total_enrolls: typeof value.total_enrolls === 'number' ? value.total_enrolls : 0,
    total_revenue: typeof value.total_revenue === 'number' ? value.total_revenue : 0,
    total_views: typeof value.total_views === 'number' ? value.total_views : 0,
    avg_rating: typeof value.avg_rating === 'number' ? value.avg_rating : 0,
  };
}

function parseCategoryDetail(payload: unknown): AdminCategoryDetail {
  if (!isRecord(payload)) {
    throw new Error('INVALID_CATEGORY_DETAIL_PAYLOAD');
  }

  return {
    ...parseCategory(payload),
    translations_full: Array.isArray(payload.translations_full)
      ? payload.translations_full.map(parseTranslationFull)
      : [],
    stats: parseDetailStats(payload.stats),
    courses: Array.isArray(payload.courses) ? payload.courses.map(parseCourseSummary) : [],
    generated_at: typeof payload.generated_at === 'string' ? payload.generated_at : '',
  };
}

function parseCategoryListResponse(payload: unknown): AdminCategoryListResponse {
  if (!isRecord(payload) || !Array.isArray(payload.items)) {
    throw new Error('INVALID_CATEGORY_LIST_PAYLOAD');
  }

  return {
    items: payload.items.map(parseCategory),
    total: typeof payload.total === 'number' ? payload.total : 0,
    page: typeof payload.page === 'number' ? payload.page : 1,
    page_size: typeof payload.page_size === 'number' ? payload.page_size : 20,
    total_pages: typeof payload.total_pages === 'number' ? payload.total_pages : 1,
    has_next: typeof payload.has_next === 'boolean' ? payload.has_next : false,
    has_prev: typeof payload.has_prev === 'boolean' ? payload.has_prev : false,
  };
}

function parseCategoryStats(value: unknown): AdminCategoryStats {
  if (!isRecord(value)) {
    throw new Error('INVALID_CATEGORY_STATS_PAYLOAD');
  }

  return {
    category_id: typeof value.category_id === 'string' ? value.category_id : '',
    total_courses: typeof value.total_courses === 'number' ? value.total_courses : 0,
    total_enrolls: typeof value.total_enrolls === 'number' ? value.total_enrolls : 0,
    total_revenue: typeof value.total_revenue === 'number' ? value.total_revenue : 0,
  };
}

function parseCategoryStatsResponse(payload: unknown): AdminCategoryStatsResponse {
  if (!isRecord(payload) || !Array.isArray(payload.stats)) {
    throw new Error('INVALID_CATEGORY_STATS_RESPONSE');
  }

  return {
    stats: payload.stats.map(parseCategoryStats),
    generated_at: typeof payload.generated_at === 'string' ? payload.generated_at : '',
  };
}

async function getErrorCode(response: Response, fallback: string): Promise<string> {
  try {
    const payload: unknown = await response.json();
    if (!isRecord(payload)) return fallback;
    const detail = payload.detail;
    if (isRecord(detail) && typeof detail.code === 'string') {
      return detail.code;
    }
    if (typeof payload.code === 'string') {
      return payload.code;
    }
    return fallback;
  } catch {
    return fallback;
  }
}

export const adminCategoryService = {
  async getCategories(params: ListCategoriesParams = {}): Promise<AdminCategoryListResponse> {
    const sp = new URLSearchParams();

    if (params.lang) sp.set('lang', params.lang);
    if (params.page) sp.set('page', String(params.page));
    if (params.page_size) sp.set('page_size', String(params.page_size));
    if (params.search) sp.set('search', params.search);
    if (typeof params.has_courses === 'boolean') sp.set('has_courses', String(params.has_courses));
    if (typeof params.min_course_count === 'number') sp.set('min_course_count', String(params.min_course_count));
    if (params.sort_by) sp.set('sort_by', params.sort_by);
    if (params.sort_order) sp.set('sort_order', params.sort_order);

    const suffix = sp.toString();
    const response = await apiFetch(`/categories${suffix ? `?${suffix}` : ''}`);

    if (!response.ok) {
      throw new Error('CATEGORY_LIST_REQUEST_FAILED');
    }

    return parseCategoryListResponse(await response.json());
  },

  async createCategory(payload: CreateCategoryPayload): Promise<AdminCategory> {
    const response = await apiFetch('/categories', {
      method: 'POST',
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new CategoryApiError(
        await getErrorCode(response, 'CATEGORY_CREATE_FAILED'),
        response.status,
      );
    }

    return parseCategory(await response.json());
  },

  async getStats(): Promise<AdminCategoryStatsResponse> {
    const response = await apiFetch('/categories/stats');

    if (!response.ok) {
      throw new Error('CATEGORY_STATS_REQUEST_FAILED');
    }

    return parseCategoryStatsResponse(await response.json());
  },

  async getCategoryDetail(id: string, courseLimit = 20): Promise<AdminCategoryDetail> {
    const sp = new URLSearchParams({ lang: 'vi', course_limit: String(courseLimit) });
    const response = await apiFetch(`/categories/${id}?${sp.toString()}`);

    if (!response.ok) {
      throw new CategoryApiError(
        await getErrorCode(response, 'CATEGORY_DETAIL_FAILED'),
        response.status,
      );
    }

    return parseCategoryDetail(await response.json());
  },

  async updateCategory(id: string, payload: UpdateCategoryPayload, courseLimit = 20): Promise<AdminCategoryDetail> {
    const sp = new URLSearchParams({ lang: 'vi', course_limit: String(courseLimit) });
    const response = await apiFetch(`/categories/${id}?${sp.toString()}`, {
      method: 'PATCH',
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new CategoryApiError(
        await getErrorCode(response, 'CATEGORY_UPDATE_FAILED'),
        response.status,
      );
    }

    return parseCategoryDetail(await response.json());
  },

  async deleteCategory(id: string): Promise<void> {
    const response = await apiFetch(`/categories/${id}`, {
      method: 'DELETE',
    });

    if (!response.ok) {
      throw new CategoryApiError(
        await getErrorCode(response, 'CATEGORY_DELETE_FAILED'),
        response.status,
      );
    }
  },
};

File: backend/app/api/v1/categories.py
"""
Category API routes - GET endpoints với pagination, filter, sort.
"""
from __future__ import annotations

import math
from typing import Annotated, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.core.error_codes import ErrorCodes
from app.core.dependencies import DBSession
from app.core.dependencies import RequireRole
from app.schemas.category import (
    CategoryCreateFromVietnamese,
    CategoryDetail,
    CategoryDetailStats,
    CategoryFilters,
    CategoryFullDetail,
    CategoryListResponse,
    CategoryCourseSummary,
    CategoryCourseTranslationSummary,
    CategoryRead,
    CategoryStats,
    CategoryStatsResponse,
    CategoryTranslationSummary,
    CategoryUpdateFromVietnamese,
)
from app.services.category import CategoryService

router = APIRouter(prefix="/categories", tags=["categories"])
AdminUser = Annotated[UUID, Depends(RequireRole(["admin"]))]


def _build_translations_dict(category, lang: str = "vi") -> dict[str, CategoryTranslationSummary]:
    """Build translations dict từ category object."""
    translations = {}
    for t in category.category_translations:
        translations[t.lang] = CategoryTranslationSummary(
            lang=t.lang,
            name=t.name or "",
            description=t.description,
        )
    return translations


def _get_course_count(category, course_counts: dict) -> int:
    """Get course count cho category."""
    return course_counts.get(category.id, 0)


def _build_category_detail(category, course_counts: dict) -> CategoryDetail:
    from app.schemas.category import CategoryTranslationRead

    return CategoryDetail(
        id=category.id,
        icon=category.icon,
        created_at=category.created_at,
        updated_at=category.updated_at,
        translations=_build_translations_dict(category),
        course_count=_get_course_count(category, course_counts),
        translations_full=[
            CategoryTranslationRead(
                id=t.id,
                category_id=t.category_id,
                lang=t.lang,
                name=t.name or "",
                description=t.description,
                is_auto_translated=t.is_auto_translated or False,
                created_at=t.created_at or category.created_at,
                updated_at=t.updated_at or category.updated_at,
            )
            for t in category.category_translations
        ],
    )


def _build_course_translations_dict(course) -> dict[str, CategoryCourseTranslationSummary]:
    """Build compact course translation map."""
    translations = {}
    for translation in course.course_translations:
        translations[translation.lang] = CategoryCourseTranslationSummary(
            lang=translation.lang,
            title=translation.title or "",
            subtitle=translation.subtitle,
            description=translation.description,
            slug=translation.slug,
        )
    return translations


def _build_course_summary(course) -> CategoryCourseSummary:
    """Build embedded course summary for category detail."""
    return CategoryCourseSummary(
        id=course.id,
        default_slug=course.default_slug,
        thumbnail_url=course.thumbnail_url,
        level=getattr(course.level, "value", str(course.level)),
        is_published=course.is_published,
        approval_status=course.approval_status,
        base_price=course.base_price or 0,
        currency=course.currency,
        estimated_duration=course.estimated_duration,
        lessons_count=course.lessons_count or 0,
        views=course.views or 0,
        total_enrolls=course.total_enrolls or 0,
        revenue=course.revenue or 0,
        rating_avg=float(course.rating_avg or 0),
        created_at=course.created_at,
        updated_at=course.updated_at,
        translations=_build_course_translations_dict(course),
    )


def _build_category_full_detail(
    category,
    course_counts: dict,
    stats: dict[str, int | float],
    courses: list,
) -> CategoryFullDetail:
    from datetime import datetime
    from app.schemas.category import CategoryTranslationRead

    return CategoryFullDetail(
        id=category.id,
        icon=category.icon,
        created_at=category.created_at,
        updated_at=category.updated_at,
        translations=_build_translations_dict(category),
        course_count=_get_course_count(category, course_counts),
        translations_full=[
            CategoryTranslationRead(
                id=t.id,
                category_id=t.category_id,
                lang=t.lang,
                name=t.name or "",
                description=t.description,
                is_auto_translated=t.is_auto_translated or False,
                created_at=t.created_at or category.created_at,
                updated_at=t.updated_at or category.updated_at,
            )
            for t in category.category_translations
        ],
        stats=CategoryDetailStats(
            total_courses=int(stats.get("total_courses", 0)),
            published_courses=int(stats.get("published_courses", 0)),
            draft_courses=int(stats.get("draft_courses", 0)),
            total_enrolls=int(stats.get("total_enrolls", 0)),
            total_revenue=int(stats.get("total_revenue", 0)),
            total_views=int(stats.get("total_views", 0)),
            avg_rating=float(stats.get("avg_rating", 0)),
        ),
        courses=[_build_course_summary(course) for course in courses],
        generated_at=datetime.utcnow(),
    )


@router.post(
    "",
    response_model=CategoryDetail,
    status_code=status.HTTP_201_CREATED,
    summary="Create category from Vietnamese input",
    description="Admin nhập name/description tiếng Việt, AI tự sinh bản dịch tiếng Anh",
)
async def create_category(
    payload: CategoryCreateFromVietnamese,
    db: DBSession,
    _: AdminUser,
) -> CategoryDetail:
    """
    Create category with required Vietnamese translation and generated English translation.
    """
    service = CategoryService(db)
    category = await service.create_category_from_vietnamese(payload)
    course_counts = await service.get_course_counts([category.id])
    return _build_category_detail(category, course_counts)


@router.patch(
    "/{category_id}",
    response_model=CategoryFullDetail,
    summary="Update category Vietnamese content",
    description="Admin chỉ sửa tiếng Việt; nếu nội dung đổi, AI tự sinh lại bản tiếng Anh",
)
async def update_category(
    category_id: UUID,
    payload: CategoryUpdateFromVietnamese,
    db: DBSession,
    _: AdminUser,
    lang: Annotated[str, Query(description="Primary language")] = "vi",
    course_limit: Annotated[
        int,
        Query(ge=0, le=100, description="Maximum related courses to include"),
    ] = 20,
) -> CategoryFullDetail:
    """Patch Vietnamese category content and return the full category resource."""
    service = CategoryService(db)
    category = await service.update_category_from_vietnamese(category_id, payload)
    course_counts = await service.get_course_counts([category.id])
    stats = await service.get_category_detail_stats(category.id)
    courses = await service.get_category_courses(category.id, limit=course_limit)
    return _build_category_full_detail(category, course_counts, stats, courses)


@router.delete(
    "/{category_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete category",
    description="Xoa category khi category chua co khoa hoc nao lien ket. Tra ve 409 neu co khoa hoc.",
)
async def delete_category(
    category_id: UUID,
    db: DBSession,
    _: AdminUser,
) -> None:
    """Xoa category khi khong co khoa hoc lien ket."""
    try:
        service = CategoryService(db)
        await service.delete_category(category_id)
    except HTTPException:
        raise
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={
                "code": ErrorCodes.CATEGORY_DELETE_FAILED,
                "message": "Could not delete category.",
            },
        )


@router.get(
    "",
    response_model=CategoryListResponse,
    summary="Get all categories",
    description="Paginated list với filter, sort, search, course count",
)
async def get_categories(
    db: DBSession,
    lang: Annotated[
        str,
        Query(description="Language code (vi|en|jp...)", examples=["vi", "en"])
    ] = "vi",
    page: Annotated[
        int,
        Query(ge=1, description="Page number")
    ] = 1,
    page_size: Annotated[
        int,
        Query(ge=1, le=100, description="Items per page")
    ] = 20,
    search: Annotated[
        Optional[str],
        Query(description="Search in category name")
    ] = None,
    has_courses: Annotated[
        Optional[bool],
        Query(description="Filter: only categories with/without courses")
    ] = None,
    min_course_count: Annotated[
        Optional[int],
        Query(ge=0, description="Minimum course count")
    ] = None,
    sort_by: Annotated[
        str,
        Query(description="Sort by: created_at|updated_at|name")
    ] = "created_at",
    sort_order: Annotated[
        str,
        Query(pattern="^(asc|desc)$", description="Sort direction")
    ] = "desc",
) -> CategoryListResponse:
    """
    Get paginated categories với đầy đủ tính năng:
    
    - **page/page_size**: Phân trang
    - **lang**: Ngôn ngữ cho translations
    - **search**: Tìm kiếm theo tên
    - **has_courses**: Lọc categories có/không có khóa học
    - **min_course_count**: Lọc theo số khóa học tối thiểu
    - **sort_by/sort_order**: Sắp xếp
    """
    filters = CategoryFilters(
        lang=lang,
        search=search,
        has_courses=has_courses,
        min_course_count=min_course_count,
        sort_by=sort_by,
        sort_order=sort_order,
    )

    service = CategoryService(db)
    categories, total = await service.get_categories(
        page=page,
        page_size=page_size,
        filters=filters,
    )

    # Get course counts for the categories on this page.
    course_counts = await service.get_course_counts([cat.id for cat in categories])

    # Build response
    items = [
        CategoryRead(
            id=cat.id,
            icon=cat.icon,
            created_at=cat.created_at,
            updated_at=cat.updated_at,
            translations=_build_translations_dict(cat, lang),
            course_count=_get_course_count(cat, course_counts),
        )
        for cat in categories
    ]

    total_pages = math.ceil(total / page_size) if total > 0 else 1

    return CategoryListResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=total_pages,
        has_next=page < total_pages,
        has_prev=page > 1,
    )


@router.get(
    "/all",
    response_model=list[CategoryRead],
    summary="Get all categories (no pagination)",
    description="Lấy tất cả categories, không phân trang (cho dropdown/cache)",
)
async def get_all_categories(
    db: DBSession,
    lang: Annotated[str, Query(description="Language code")] = "vi",
    include_stats: Annotated[
        bool,
        Query(description="Include course counts")
    ] = False,
) -> list[CategoryRead]:
    """
    Get all categories without pagination.
    Phù hợp cho dropdown lists hoặc caching.
    """
    service = CategoryService(db)
    categories = await service.get_all_categories(lang=lang, include_stats=include_stats)

    course_counts = await service.get_course_counts([cat.id for cat in categories])

    return [
        CategoryRead(
            id=cat.id,
            icon=cat.icon,
            created_at=cat.created_at,
            updated_at=cat.updated_at,
            translations=_build_translations_dict(cat, lang),
            course_count=_get_course_count(cat, course_counts),
        )
        for cat in categories
    ]


@router.get(
    "/stats",
    response_model=CategoryStatsResponse,
    summary="Get category statistics",
    description="Thống kê số khóa học, enrollments, revenue cho tất cả categories",
)
async def get_category_stats(
    db: DBSession,
) -> CategoryStatsResponse:
    """Get statistics cho all categories."""
    from datetime import datetime

    service = CategoryService(db)
    stats = await service.get_category_stats()

    category_stats = [
        CategoryStats(
            category_id=cat_id,
            total_courses=data.get("total_courses", 0),
            total_enrolls=data.get("total_enrolls", 0),
            total_revenue=data.get("total_revenue", 0),
        )
        for cat_id, data in stats.items()
    ]

    return CategoryStatsResponse(
        stats=category_stats,
        generated_at=datetime.utcnow(),
    )


@router.get(
    "/{category_id}",
    response_model=CategoryFullDetail,
    summary="Get category by ID",
    description="Lấy chi tiết một category với translations, thống kê và khóa học liên quan",
)
async def get_category_by_id(
    category_id: UUID,
    db: DBSession,
    lang: Annotated[str, Query(description="Primary language")] = "vi",
    course_limit: Annotated[
        int,
        Query(ge=0, le=100, description="Maximum related courses to include"),
    ] = 20,
) -> CategoryFullDetail:
    """Get complete category detail by UUID."""
    try:
        service = CategoryService(db)
        category = await service.get_category_by_id(category_id, lang)

        if not category:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail={
                    "code": ErrorCodes.CATEGORY_NOT_FOUND,
                    "message": f"Category {category_id} not found",
                },
            )

        course_counts = await service.get_course_counts([category.id])
        stats = await service.get_category_detail_stats(category.id)
        courses = await service.get_category_courses(category.id, limit=course_limit)

        return _build_category_full_detail(category, course_counts, stats, courses)
    except HTTPException:
        raise
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={
                "code": ErrorCodes.CATEGORY_DETAIL_FAILED,
                "message": "Could not load category detail.",
            },
        )
